File: module/service-grpc/service_watch.go
package main

import (
	"encoding/binary"
	"errors"
	"sync"

	"github.com/pantopic/ext-buffer/sdk-go"
	"github.com/pantopic/ext-grpc-server/sdk-go"

	internal "github.com/pantopic/turbokube/module/service-grpc/internal"
)

var evtPool = sync.Pool{New: func() any { return &internal.Event{} }}
var watchEventBatch = &internal.WatchEventBatch{Event: &internal.Event{}}
var watchEventSync = &internal.WatchEventSync{}
var watch_buffer = make([]byte, PCB_RESPONSE_SIZE_MAX())

func shardRecv(_, data []byte, id uint64) {
	var err error
	if id == WATCH_ID_ERROR {
		println(`watch err ` + string(data))
		err = errors.New(string(data))
		return
	}
	switch data[0] {
	case WatchMessageType_INIT:
		watchResp.Reset()
		respHeader.Reset()
		if err = respHeader.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal response header: ` + err.Error())
		}
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		watchResp.Created = true
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	case WatchMessageType_EVENT_BATCH:
		watchEventBatch.Reset()
		if err = watchEventBatch.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal watch event batch: ` + err.Error())
		}
		if len(watchEventBatch.WatchIdsPrev) > 0 {
			b, err := watchEventBatch.Event.MarshalVT()
			if err != nil {
				panic(`Unable to marshal watch event: ` + err.Error())
			}
			for _, id := range watchEventBatch.WatchIdsPrev {
				sendEvent(id, watchEventBatch.Revision, b)
			}
		}
		if len(watchEventBatch.WatchIds) > 0 {
			watchEventBatch.Event.PrevKv = nil
			b, err := watchEventBatch.Event.MarshalVT()
			if err != nil {
				panic(`Unable to marshal watch event: ` + err.Error())
			}
			for _, id := range watchEventBatch.WatchIds {
				sendEvent(id, watchEventBatch.Revision, b)
			}
		}
	case WatchMessageType_EVENT_SYNC:
		watchEventSync.Reset()
		if err = watchEventSync.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal watch event batch: ` + err.Error())
		}
		for _, id := range watchEventSync.IDs {
			events := bufferPoolWatchEvent.Find(uint64(id))
			clearEvents(events, id, watchEventSync.Revision, true)
		}
	case WatchMessageType_NOTIFY:
		watchResp.Reset()
		respHeader.Reset()
		if err = respHeader.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal response header: ` + err.Error())
		}
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	case WatchMessageType_CANCELED:
		watchResp.Reset()
		watchResp.WatchId = int64(id)
		watchResp.Canceled = true
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
		// TODO: reset watch buffer pool
	case WatchMessageType_ERR_EXISTS:
		watchResp.Reset()
		watchResp.WatchId = -1
		watchResp.Created = true
		watchResp.Canceled = true
		watchResp.CancelReason = ErrWatcherDuplicateID.Error()
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	case WatchMessageType_ERR_COMPACTED:
		respHeader.Reset()
		if err = respHeader.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal response header: ` + err.Error())
		}
		watchResp.Reset()
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		watchResp.Created = true
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
		watchResp.Reset()
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		watchResp.Canceled = true
		watchResp.CompactRevision = respHeader.Revision
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	default:
		panic(`Unrecognized`)
	}
}

func sendEvent(id int64, rev uint64, b []byte) {
	events := bufferPoolWatchEvent.Find(uint64(id))
	b2 := binary.BigEndian.AppendUint64(b, rev)
	if events.Append(b2) {
		return
	}
	clearEvents(events, id, rev, false)
	if !events.Append(b2) {
		panic(`Failed to append watch event after reset`)
	}
}

func clearEvents(events buffer.MultiValue, id int64, rev uint64, sync bool) {
	var lastRev uint64
	respHeader.Reset()
	watchResp.Reset()
	watchResp.Header = respHeader
	watchResp.WatchId = int64(id)
	for b := range events.Iter() {
		evt := evtPool.Get().(*internal.Event)
		if err := evt.UnmarshalVT(b[:len(b)-8]); err != nil {
			panic(`Unable to unmarshal event B: ` + err.Error())
		}
		lastRev = binary.BigEndian.Uint64(b[len(b)-8:])
		watchResp.Events = append(watchResp.Events, evt)
	}
	if len(watchResp.Events) == 0 {
		return
	}
	watchResp.Fragment = !sync && lastRev == rev
	watchResp.Header.Revision = int64(rev)
	res, err := watchResp.MarshalVT()
	if err != nil {
		panic(`Unable to marshal watch response: ` + err.Error())
	}
	grpc_server.Send(res)
	events.Reset()
	for _, evt := range watchResp.Events {
		evt.Reset()
		evtPool.Put(evt)
	}
}

var watchResp = &internal.WatchResponse{
	Header: &internal.ResponseHeader{},
}
var respHeader = &internal.ResponseHeader{}

func watchOpen() (err error) {
	return kvShard().StreamOpen([]byte(`watch`))
}

func watchRecv(data []byte) (err error) {
	return kvShard().StreamSend([]byte(`watch`), data)
}

func watchClose() (err error) {
	return
}

File: module-zig/service-grpc/service_watch.zig
//! Mirrors module/service-grpc/service_watch.go

const std = @import("std");
const pb = @import("pb/etcdserverpb.pb.zig");
const buffer = @import("buffer");

const errors = @import("error.zig");
const module = @import("module.zig");
const types = @import("types.zig");
const util = @import("util.zig");

var arena_state = std.heap.ArenaAllocator.init(std.heap.wasm_allocator);
var out: [1536 * 1024 + 8]u8 = undefined;

fn decode(comptime T: type, b: []const u8) !T {
    var reader = std.Io.Reader.fixed(b);
    return T.decode(&reader, arena_state.allocator());
}

fn encode(msg: anytype) ![]const u8 {
    var writer = std.Io.Writer.fixed(&out);
    try msg.encode(&writer, arena_state.allocator());
    return writer.buffered();
}

pub fn shardRecv(name: []const u8, data: []const u8, id: u64) void {
    _ = name;
    if (id == types.WATCH_ID_ERROR) {
        std.debug.print("watch err {s}\n", .{data});
        return;
    }
    defer _ = arena_state.reset(.retain_capacity);
    switch (data[0]) {
        types.WatchMessageType_INIT => {
            const header = decode(pb.ResponseHeader, data[1..]) catch |err| {
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            };
            const resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
                .created = true,
            };
            const b = encode(resp) catch |err| {
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            };
            module.server.send(b) catch unreachable;
        },
        types.WatchMessageType_EVENT_BATCH => {
            var watchEventBatch = decode(pb.WatchEventBatch, data[1..]) catch |err| {
                std.debug.panic("Unable to unmarshal watch event batch: {s}", .{@errorName(err)});
            };
            if (watchEventBatch.watch_ids_prev.items.len > 0) {
                const b = encode(watchEventBatch.event.?) catch |err| {
                    std.debug.panic("Unable to marshal watch event batch event: {s}", .{@errorName(err)});
                };
                for (watchEventBatch.watch_ids_prev.items) |watch_id| {
                    sendEvent(watch_id, watchEventBatch.revision, b) catch |err| {
                        std.debug.panic("Unable to send watch event batch event: {s}", .{@errorName(err)});
                    };
                }
            }
            if (watchEventBatch.watch_ids.items.len > 0) {
                watchEventBatch.event.?.prev_kv = null;
                const b = encode(watchEventBatch.event.?) catch |err| {
                    std.debug.panic("Unable to marshal watch event batch event: {s}", .{@errorName(err)});
                };
                for (watchEventBatch.watch_ids.items) |watch_id| {
                    sendEvent(watch_id, watchEventBatch.revision, b) catch |err| {
                        std.debug.panic("Unable to send watch event batch event: {s}", .{@errorName(err)});
                    };
                }
            }
        },
        types.WatchMessageType_EVENT_SYNC => {
            const watchEventSync = decode(pb.WatchEventSync, data[1..]) catch |err| {
                std.debug.panic("Unable to unmarshal watch event sync: {s}", .{@errorName(err)});
            };
            for (watchEventSync.IDs.items) |watch_id| {
                const events = module.bufferPoolWatchEvent.find(@bitCast(watch_id));
                clearEvents(events, watch_id, watchEventSync.revision, true) catch |err| {
                    std.debug.panic("Unable to clear watch events: {s}", .{@errorName(err)});
                };
            }
        },
        types.WatchMessageType_NOTIFY => {
            const header = decode(pb.ResponseHeader, data[1..]) catch |err| {
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            };
            const resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
            };
            const b = encode(resp) catch |err| {
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            };
            module.server.send(b) catch {};
        },
        types.WatchMessageType_CANCELED => {
            const resp = pb.WatchResponse{
                .watch_id = @bitCast(id),
                .canceled = true,
            };
            const b = encode(resp) catch |err| {
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            };
            module.server.send(b) catch {};
            // TODO: reset watch buffer pool
        },
        types.WatchMessageType_ERR_EXISTS => {
            const resp = pb.WatchResponse{
                .watch_id = -1,
                .created = true,
                .canceled = true,
                .cancel_reason = errors.err_watcher_duplicate_id,
            };
            const b = encode(resp) catch |err| {
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            };
            module.server.send(b) catch {};
        },
        types.WatchMessageType_ERR_COMPACTED => {
            const header = decode(pb.ResponseHeader, data[1..]) catch |err| {
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            };
            var resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
                .created = true,
            };
            var b = encode(resp) catch |err| {
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            };
            module.server.send(b) catch {};
            resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
                .canceled = true,
                .compact_revision = header.revision,
            };
            b = encode(resp) catch |err| {
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            };
            module.server.send(b) catch {};
        },
        else => @panic("Unrecognized"),
    }
}

fn sendEvent(watch_id: i64, rev: u64, data: []const u8) !void {
    var rev_buf: [8]u8 = undefined;
    std.mem.writeInt(u64, &rev_buf, rev, .big);
    const data2: []const u8 = try std.mem.concat(arena_state.allocator(), u8, &.{ data, &rev_buf });
    const events = module.bufferPoolWatchEvent.find(@bitCast(watch_id));
    if (events.append(data2)) {
        return;
    }
    clearEvents(events, watch_id, rev, false) catch |err| {
        std.debug.panic("Unable to clear watch events: {s}", .{@errorName(err)});
    };
    if (!events.append(data2)) {
        @panic("Failed to append watch event after reset");
    }
}

fn clearEvents(events: buffer.MultiValue, watch_id: i64, rev: u64, sync: bool) !void {
    var resp = pb.WatchResponse{
        .header = .{},
        .watch_id = @bitCast(watch_id),
    };
    var last_rev: u64 = 0;
    var it = events.iterator();
    while (it.next()) |b| {
        const rev_bytes: *const [8]u8 = @ptrCast(b[b.len - 8 ..].ptr);
        last_rev = std.mem.readInt(u64, rev_bytes, .big);
        const evt = decode(pb.Event, b[0 .. b.len - 8]) catch |err|
            std.debug.panic("Unable to unmarshal event: {s}", .{@errorName(err)});
        resp.events.append(arena_state.allocator(), evt) catch |err|
            std.debug.panic("{s}", .{@errorName(err)});
    }
    if (resp.events.items.len == 0) {
        return;
    }
    resp.fragment = !sync and last_rev == rev;
    resp.header.?.revision = @bitCast(last_rev);
    const res = encode(resp) catch |err|
        std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
    module.server.send(res) catch {};
    events.reset();
}

pub fn open() anyerror!void {
    return util.kvShard().StreamOpen("watch");
}

pub fn recv(data: []const u8) anyerror!void {
    return util.kvShard().StreamSend("watch", data);
}

pub fn close() anyerror!void {}
